src/_griffe/docstrings/google.py

```diff
@@ -491,22 +552,18 @@ def _read_returns_section(
                     annotation = docstring.parent.annotation  # type: ignore[union-attr]
                 else:
                     raise ValueError
-                if len(block) > 1:
-                    if annotation.is_tuple:
-                        annotation = annotation.slice.elements[index]
-                    else:
-                        if annotation.is_iterator:
-                            return_item = annotation.slice
-                        elif annotation.is_generator:
-                            return_item = annotation.slice.elements[2]
-                        else:
-                            raise ValueError
-                        if isinstance(return_item, ExprName):
-                            annotation = return_item
-                        elif return_item.is_tuple:
-                            annotation = return_item.slice.elements[index]
-                        else:
-                            annotation = return_item
+                orig_annotation = annotation
+                annotation = _maybe_destructure_annotation(
+                    annotation,
+                    index,
+                    multiple=returns_multiple_items,
+                )
+                if annotation == orig_annotation:
+                    annotation = _maybe_destructure_annotation(
+                        _unpack_generators(annotation, 2),
+                        index,
+                        multiple=returns_multiple_items,
+                    )
```

Comment:
Tested locally and tests pass with this change:

```suggestion
                annotation = _maybe_destructure_annotation(
                    _unpack_generators(annotation, 2),
                    index,
                    multiple=returns_multiple_items,
                )
```

I wonder if we're missing something though, but IIRC tests are quite comprehensive?

Reply:
It's your call. I translated the code very directly from the non-PR version, and it included this try-unpacking-else-un-generator-then-unpack dance. I *did* get a "organically grown and never pruned" vibe from this part, but I didn't study Griffe's workings and code history deeply enough to be comfortable with pruning this section… at least not without double-checking with you first.

In any case, I don't see any problems with your suggestion, based solely on reading and rewriting that code section.

Reply:
Thanks, that makes sense.

> an "organically grown and never pruned" vibe

Definitely. The parsers were never something I felt interested in (for many reasons), and I just coded them lazily, in an imperative mood. I didn't spend much time trying to make them elegant or efficient, as long as they are working as intended. Our tests suite says the suggestion works, so lets apply it. If it turns out the tests are insufficient, we'll fix that then :relaxed: